### PR TITLE
Add `--reuse` flag for `publish` command

### PR DESCRIPTION
**Pull request by presztak**
- Fail the publish by default if the alias already exists
- Add `--reuse` flag as a way to override that which will lead to the current image being deleted and replaced

Fixes: #11489

### CODE REVIEW

File: lxc/utils.go [hunk 1]
```diff
@@ -109,6 +109,44 @@ func parseDeviceOverrides(deviceOverrideArgs []string) (map[string]map[string]st
...
+	if len(aliases) == 0 {
+		return nil, nil
+	}
+
+	names := make([]string, len(aliases))
+	for i, alias := range aliases {
+		names[i] = alias.Name
+	}
+
+	sort.Strings(names)
```
Comment (tomponline) [inline lxc/utils.go:140]: Why do we sort them?

File: lxc/publish.go [hunk 2]
```diff
@@ -227,6 +229,52 @@ func (c *cmdPublish) Run(cmd *cobra.Command, args []string) error {
...
+			if image != nil {
+				_, found := visitedImages[image.Fingerprint]
+				if found {
+					continue
+				}
+
+				visitedImages[image.Fingerprint] = nil
+			}
+
+			// An image can have multiple aliases. An image should be removed
```
Comment (tomponline) [inline lxc/publish.go:261]: Please can you explain this logic further, im not quite following the reasoning here. Thanks

File: lxc/utils.go (hunk 1 above)
Comment (presztak) [reply lxc/utils.go:141]: `GetExistingAliases` which is using `sort.SearchStrings` requires sorted slice. I've added comment.

File: lxc/utils.go [hunk 3]
```diff
@@ -109,6 +109,44 @@ func parseDeviceOverrides(deviceOverrideArgs []string) (map[string]map[string]st
...
+		if !found {
+			return false
+		}
+	}
+
+	return true
+}
+
+// GetCommonAliases returns the common aliases between a list of aliases and all the existing ones.
+func GetCommonAliases(client lxd.InstanceServer, aliases []api.ImageAlias) ([]api.ImageAliasesEntry, error) {
```
[the remark this answers is not in the document]
Comment (presztak) [reply lxc/utils.go]: Done.

File: lxc/publish.go (hunk 2 above)
Comment (tomponline) [reply lxc/publish.go:261]: Thanks for the explanation, that makes sense now.

Comment (tomponline) [reply lxc/publish.go:261]: So if an image being published reuses all the aliases from an existing image then that existing image is removed.

Comment (tomponline) [reply lxc/publish.go:261]: Could you add a comment that to effect for future readers please?

File: test/suites/clustering.sh [hunk 4]
```diff
@@ -2756,7 +2756,7 @@ test_clustering_image_refresh() {
   # Modify public testimage
   old_fingerprint="$(LXD_DIR="${LXD_REMOTE_DIR}" lxc image ls testimage -c f --format csv)"
   dd if=/dev/urandom count=32 | LXD_DIR="${LXD_REMOTE_DIR}" lxc file push - c1/foo
-  LXD_DIR="${LXD_REMOTE_DIR}" lxc publish c1 --alias testimage --public
+  LXD_DIR="${LXD_REMOTE_DIR}" lxc publish c1 --alias testimage --reuse --public
```
Comment (tomponline) [inline test/suites/clustering.sh:2759]: Did you add this because the old behaviour was to reuse by default?

**Comment by tomponline** [comment]
Just a couple of small suggestions then I think this is good to go. Thanks!

File: test/suites/clustering.sh (hunk 4 above)
Comment (presztak) [reply test/suites/clustering.sh:2759]: Yes.

File: test/suites/basic.sh [hunk 5]
```diff
@@ -151,8 +151,10 @@ test_basic_usage() {
   # change the container filesystem so the resulting image is different
   lxc exec baz touch /somefile
   lxc stop baz --force
-  # publishing another image with same alias doesn't fail
-  lxc publish baz --alias=foo-image
+  # publishing another image with same alias fail
+  ! lxc publish baz --alias=foo-image || false
+  # publishing another image with same alias and reuse flag
+  lxc publish baz --alias=foo-image --reuse
```
[the remark this answers is not in the document]
Comment (presztak) [reply test/suites/basic.sh:157]: Sure! Will do.

**Comment by tomponline** [comment]
jenkins: test this please

File: lxc/publish.go (hunk 2 above)
Comment (presztak) [reply lxc/publish.go]: Done.

File: test/suites/basic.sh (hunk 5 above)
[the remark this answers is not in the document]
Comment (presztak) [reply test/suites/basic.sh:157]: Done.

**Comment by tomponline** [comment]
jenkins: test this please

**Review by tomponline — APPROVED** [review]
Thanks!

**Pull request merged**